aptos-move/framework/aptos-framework/sources/vesting.move

```diff
@@ -0,0 +1,1239 @@
+        assert!(period_duration > 0, error::invalid_argument(EZERO_VESTING_SCHEDULE_PERIOD));
+        assert!(
+            start_timestamp_secs >= timestamp::now_seconds(),
+            error::invalid_argument(EVESTING_START_TOO_SOON),
+        );
+
+        VestingSchedule {
+            schedule,
+            start_timestamp_secs,
+            period_duration,
+            last_vested_period: 0,
+        }
+    }
+
+    /// Create a vesting contract with a given configurations.
+    public fun create_vesting_contract(
+        admin: &signer,
+        grant: Coin<AptosCoin>,
+        shareholders: &vector<address>,
+        shares: &vector<u64>,
+        vesting_schedule: VestingSchedule,
+        withdrawal_address: address,
+        operator: address,
+        voter: address,
+        commission_percentage: u64,
+        // Optional seed used when creating the staking contract account.
+        contract_creation_seed: vector<u8>,
+    ): address acquires AdminStore {
+        assert!(coin::value(&grant) > 0, error::invalid_argument(EZERO_GRANT));
+        assert!(
+            !system_addresses::is_framework_reserved_address(withdrawal_address) &&
+            !system_addresses::is_vm_address(withdrawal_address),
+            error::invalid_argument(EINVALID_WITHDRAWAL_ADDRESS),
+        );
+        assert_account_is_registered_for_apt(withdrawal_address);
+        assert!(vector::length(shareholders) > 0, error::invalid_argument(ENO_SHAREHOLDERS));
+        assert!(
+            vector::length(shares) == vector::length(shareholders),
+            error::invalid_argument(ESHARES_LENGTH_MISMATCH),
+        );
+
+        // Create a coins pool to track shareholders and shares of the grant.
+        let grant_amount = coin::value(&grant);
+        let grant_pool = pool_u64::create(MAXIMUM_SHAREHOLDERS);
+        let len = vector::length(shareholders);
+        let i = 0;
+        while (i < len) {
+            pool_u64::buy_in(
+                &mut grant_pool,
+                *vector::borrow(shareholders, i),
+                *vector::borrow(shares, i),
+            );
+            i = i + 1;
+        };
+        pool_u64::update_total_coins(&mut grant_pool, grant_amount);
+
+        // If this is the first time this admin account has created a vesting contract, initialize the admin store.
+        let admin_address = signer::address_of(admin);
+        if (!exists<AdminStore>(admin_address)) {
+            move_to(admin, AdminStore {
+                vesting_contracts: vector::empty<address>(),
+                nonce: 0,
+                create_events: new_event_handle<CreateVestingContractEvent>(admin),
+            });
+        };
+
+        // Initialize the vesting contract in a new resource account. This allows the same admin to create multiple
+        // pools.
+        let (contract_signer, contract_signer_cap) = create_vesting_contract_account(admin, contract_creation_seed);
```

Comment:
I still try understand the overview of this contract. what is the main purpose of having this resource account?

Reply:
It isolates the power to control the staking. If we don't create this account and use admin's account directly, the admin would be able to bypass this vesting module and manage the stake themselves, including arbitrarily withdrawing.